loopchain/channel/channel_service.py

```diff
@@ -122,53 +120,63 @@ async def _serve():
             logging.info(f'channel_service: init complete channel: {ChannelProperty().name}, '
                          f'state({self.__state_machine.state})')
 
-        loop = MessageQueueService.loop
-        # loop.set_debug(True)
+        loop = self.__inner_service.loop
         loop.create_task(_serve())
-        loop.add_signal_handler(signal.SIGINT, self.close)
-        loop.add_signal_handler(signal.SIGTERM, self.close)
+        loop.add_signal_handler(signal.SIGINT, self.close, signal.SIGINT)
+        loop.add_signal_handler(signal.SIGTERM, self.close, signal.SIGTERM)
 
         try:
             loop.run_forever()
         except Exception as e:
             traceback.print_exception(type(e), e, e.__traceback__)
         finally:
             loop.run_until_complete(loop.shutdown_asyncgens())
+            self._cancel_tasks(loop)
+            self._cleanup()
             loop.close()
 
-            self.cleanup()
-
-    def close(self):
+    def close(self, signum=None):
+        logging.info(f"close() signum = {repr(signum)}")
         if self.__inner_service:
             self.__inner_service.cleanup()
-            logging.info("Cleanup ChannelInnerService.")
 
-        MessageQueueService.loop.stop()
+        self.__inner_service.loop.stop()
 
-    def cleanup(self):
-        logging.info("Cleanup Channel Resources.")
+    @staticmethod
+    def _cancel_tasks(loop):
+        for task in asyncio.Task.all_tasks(loop):
+            if task.done():
+                continue
+            task.cancel()
+            try:
+                loop.run_until_complete(task)
+            except asyncio.CancelledError as e:
+                logging.info(f"_cancel_tasks() task : {task}, error : {e}")
 
-        if self.__block_manager:
-            self.__block_manager.stop()
-            self.__block_manager = None
-            logging.info("Cleanup BlockManager.")
+    def _cleanup(self):
+        logging.info("_cleanup() Channel Resources.")
+
+        if self.__timer_service.is_run():
+            self.__timer_service.stop()
+            self.__timer_service.wait()
+            logging.info("_cleanup() TimerService.")
```

Comment:
Can you explain why `TimerService` should be cleared at first between Channel Resources?

Reply:
I want to cleanup in reverse order of the start order.
start order is :
 1. broadcast_scheduler
 1. block_manager
 1. score_container
 1. sub_services
 1. timer_service

Exception cases are `block_manager` and `sub_services`.
`sub_services` must cleanup before stop `asyncio loop`.
`block_manager` must cleanup last of order cause this PR created.

So, cleanup order is :
 1. sub_services
 1. timer_service
 1. score_container
 1. broadcast_scheduler
 1. block_manager

Reply:
It looks reasonable.